rfcs/001-fdb-revision-metadata-model.md

```diff
@@ -0,0 +1,215 @@
+
+This design stores each edit branch as its own KV, and all of the edit branches
+are stored separately from the actual document data. Document reads can avoid
+retrieving this information, while writes can avoid retrieving the document
+body.
+
+## Requirements Language
+
+[NOTE]: # ( Do not alter the section below. Follow its instructions. )
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC
+2119](https://www.rfc-editor.org/rfc/rfc2119.txt).
+
+## Terminology
+
+[TIP]:  # ( Provide a list of any unique terms or acronyms, and their
+definitions here.)
+
+`Versionstamp`: a 12 byte, unique, monotonically (but not sequentially)
+increasing value for each committed transaction. The first 8 bytes are the
+committed version of the database. The next 2 bytes are monotonic in the
+serialization order for transactions. The final 2 bytes are user-defined and can
+be used to create multiple versionstamps in a single transaction.
+
+`Incarnation`: a single byte, monotonically increasing value specified for each
+CouchDB database. The `Incarnation` starts at `\x00` when a database is created
+and is incremented by one whenever a database is relocated to a different
+FoundationDB cluster.
+
+`Sequence`: a 13 byte value formed by combining the current `Incarnation` of
+the database and the `Versionstamp` of the transaction. Sequences are
+monotonically increasing even when a database is relocated across FoundationDB
+clusters.
+
+---
+
+# Detailed Description
+
+The size limits in FoundationDB preclude storing the entire revision tree as a
+single value; in pathological situations the tree could exceed 100KB. Rather, we
+propose to store each edit *branch* as a separate KV. We have two different
+value formats, one that is used for the "winning" edit branch and one used for
+any additional edit branches of the document. The winning edit branch includes
+the following information:
+
+`(“revisions”, DocID, NotDeleted, RevPosition, RevHash) = (RevFormat, Sequence,
+BranchCount, [ParentRev, GrandparentRev, …])`
+
+while the other edit branches omit the `Sequence` and `BranchCount`:
+
+`(“revisions”, DocID, NotDeleted, RevPosition, RevHash) = (RevFormat,
+[ParentRev, GrandparentRev, …])`
+
+The individual elements of the key and value are defined as follows:
+- `DocID`: the document ID
+- `NotDeleted`: `\x26` if the leaf of the edit branch is deleted, `\x27`
+  otherwise (following tuple encoding for booleans)
+- `RevPosition`: positive integer encoded using standard tuple layer encoding
+  (signed, variable-length, order-preserving)
+- `RevHash`: 16 bytes uniquely identifying this revision
+- `RevFormat`: enum for the revision encoding being used to enable schema
+  evolution
+- `Sequence`: the sequence of the last transaction that modified the document
+  (NB: not necessarily the last edit to *this* branch).
```

Comment:
Would you mind just elaborating on what you mean by `NB: not necessarily the last edit to *this* branch`?

Reply:
This is covered in the "Update Path" section, but I can summarize here. An update that extends a losing edit branch will still modify the winning edit branch KV as well, setting the Sequence (which is only stored on the winning branch) to the sequence of that transaction.

Reply:
Ok great. Thanks for clarifying.